README.md
<p align="center"><img src="https://telegra.ph/file/a0d23ed9a3fd3012ef5c2.jpg"></p>

**”It’s not who I am underneath, but what I do that defines me.” – Batman**

[![made-with-python](https://img.shields.io/badge/Made%20with-Python-1f425f.svg)](https://www.python.org/)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![API status](https://github.com/Pool-Of-Tears/SuperHero-Quotes/actions/workflows/api_status.yml/badge.svg)](https://github.com/Pool-Of-Tears/SuperHero-Quotes/actions/workflows/api_status.yml)

<p align="center"><b>An API consisting huge collection of handpicked Marvel & DC SuperHeros and Supervillains quotes!:zap:</b></p>

**Base API URL**
```
https://superhero-quotes.herokuapp.com/
```

**Click below to view endpoints**
<details>
  <summary>reveal endpoints</summary>

- `/grab?banner=mcu` or `/grab?banner=dcu`: Returns 10 random quote for the given banner. You can also fetch more than 10 quote at a time by passing `size` parameter, like: `/grab?banner=mcu&size=2`
will return two quotes from Marvel's collection.

- Note: Max size limit is `50` (fifty)

**Example response**:
```json
{
    "StatusCode": 200,
    "Banner": "Marvel Cinematic Universe (MCU)",
    "TotalQuotes": 2,
    "Items": [
        {
            "id": "6SYjJebRR8fCF9SnvFZBKQ",
            "data": {
                "author": "Iron Man",
                "quote": "You know what, give me a break, Steve. I just got hit in the head with a Hulk."
            }
        },
        {
            "id": "2cquLZiJsUbg5jtQrdwLkM",
            "data": {
                "author": "Thanos",
                "quote": "You have my respect, stark. When I'm done, half of humanity will still be alive. I hope they remember you."
            }
        }
    ]
}
```

- `/random`: Will return single randomly chosen quote from either Marvel or DC collection.

**Example response**:
```json
{
    "StatusCode": 200,
    "Banner": "DC Universe (DCU)",
    "Stuff": {
        "id": "c4ZuUeSUzjGiCEVUmsrYWN",
        "data": {
            "author": "Wonder Woman",
            "quote": "Death is necessary. It is part of life, and if we say life is a blessing, we must say that death is a blessing, as well."
        }
    }
}
```

- `/quoteId/<quote_id>` will return specific quote for the given `quote_id`. Like: `/quoteId/k3fhzAKsvCeuFhXPHPQcnT`.

**Example response**:
```json
{
    "StatusCode": 200,
    "Stuff": {
        "id": "k3fhzAKsvCeuFhXPHPQcnT",
        "data": {
            "author": "The Joker",
            "quote": "The pen, is truly mightier than the sword! "
        }
    }
}
```
</details>

## Collection updates:
Quotes collection gets updated frequently from new movie releases by both Marvel & DCU or whenever i have some great findings! Please checkout [CHANGES.md](https://github.com/starry69/SuperHero-Quotes/blob/main/CHANGES.md) file to stay in loop with new additions :)

## Contribute

Have an idea? Found a bug? Create [a new issue](https://github.com/starry69/SuperHero-Quotes/issues) with detailed description.

## License

[MIT][license] © [Stɑrry Shivɑm][github]

[license]: /LICENSE
[github]: https://github.com/starry69


CHANGES.md
# Superheroes Quotes API database updates
This file will be updated frequently whenever i add quotes from new superhero movies into API's database.

## 31st December 2020
Added quotes from DC's [Wonder Woman 1984](https://m.imdb.com/title/tt7126948/) movie.

## 20 February 2022
Added quotes from:
- [Shang-Chi and the Legend of the Ten Rings](https://m.imdb.com/title/tt9376612/)
- [The Eternals](https://m.imdb.com/title/tt9032400/)
+ [Venom: Let There Be Carnage](https://m.imdb.com/title/tt7097896/)
- [The Suicide Squad](https://m.imdb.com/title/tt6334354/)
- [Zack Snyder's Justice League](https://m.imdb.com/title/tt12361974/)


tests/test_api.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# MIT License
# Copyright (C) 2020-2021 Stɑrry Shivɑm.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
"""Includes unittests for public API endpoints."""


import pytest
import requests as r


BASE_URL = "https://superhero-quotes.herokuapp.com"


@pytest.mark.parametrize(
    ["banner", "expected_banner", "total"],
    [
        ("mcu", "Marvel Cinematic Universe (MCU)", 5),
        ("dcu", "DC Universe (DCU)", 10),
    ],
)
def test_grab_endpoint(banner, expected_banner, total):
    # make request to the API
    res = r.get(f"{BASE_URL}/grab?banner={banner}&size={total}")
    # check status code
    assert res.status_code == 200
    # check total quotes
    json_data = res.json()
    assert json_data["TotalQuotes"] == total
    assert len(json_data["Items"]) == total
    # finally check banner
    assert json_data["Banner"] == expected_banner


def test_default_grab_endpoint_size():
    # make request to the API without passing size parameter
    banners = ["dcu", "mcu"]
    for b in banners:
        res = r.get(f"{BASE_URL}/grab?banner={b}")
        # by default it should return 10 quotes
        assert res.json()["TotalQuotes"] == 10
        assert len(res.json()["Items"]) == 10


def test_invalid_grab_banner():
    # make request to API by passing 'pizza' in banner parameter
    res = r.get(f"{BASE_URL}/grab?banner=pizza")
    # status 404: not found
    assert res.status_code == 404
    assert res.json()["message"].startswith("Not found")


def test_maxsize_limit():
    # make request to API by passing more 99 in size parameter
    banners = ["dcu", "mcu"]
    size = 69
    err_msg = "You can only fetch max 50 quotes at time!"
    # make request for both banners
    for b in banners:
        res = r.get(f"{BASE_URL}/grab?banner={b}&size={size}")
        assert res.status_code == 400  # bad request
        assert res.json()["message"] == err_msg


def test_random_endpoint():
    # make request to /random endpoint.
    res = r.get(f"{BASE_URL}/random")
    # check status code
    assert res.status_code == 200
    json_data = res.json()
    assert json_data["Banner"] in {
        "DC Universe (DCU)",
        "Marvel Cinematic Universe (MCU)",
    }
    assert json_data.get("Stuff", {}).get("data") is not None


def test_quoteid_endpoint():
    # make request with specific quoteId
    quote_id = "k3fhzAKsvCeuFhXPHPQcnT"
    res = r.get(f"{BASE_URL}/quoteId/{quote_id}")
    # check status code
    assert res.status_code == 200
    json_data = res.json()
    # check if id & quote matches
    quote_text = "The pen, is truly mightier than the sword! "
    assert json_data["Stuff"]["id"] == quote_id
    assert json_data["Stuff"]["data"]["quote"] == quote_text


def test_invalid_quoteid():
    # make request with invalid quoteid
    invalid_id = "1234abcd"
    res = r.get(f"{BASE_URL}/quoteId/{invalid_id}")
    # status 404: not found
    assert res.status_code == 404
    # check error message
    err_msg = "Can't find any quote for this id."
    assert res.json()["message"] == err_msg


QuotesAPI/database/__init__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# MIT License
#
# Copyright (C) 2020-2021 Stɑrry Shivɑm
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


QuotesAPI/database/quotes_sql.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# MIT License
#
# Copyright (C) 2020-2021 Stɑrry Shivɑm
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


import shortuuid
import random

from threading import RLock
from QuotesAPI import DB_URI

from sqlalchemy import Column, String, UnicodeText
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, scoped_session


# SQLAlchemy declarative base obj
Base = declarative_base()


class MCUQuotes(Base):
    __tablename__ = "mcu_quotes"
    id = Column(String, primary_key=True)
    character = Column(UnicodeText, nullable=False)
    quote = Column(UnicodeText, nullable=False)

    def __init__(self, id, character, quote):
        self.id = id
        self.character = character
        self.quote = quote

    def __repr__(self):
        return f"id={self.ud}, character={self.character}, quote={self.quote}"


class DCUQuotes(Base):
    __tablename__ = "dcu_quotes"
    id = Column(String, primary_key=True)
    character = Column(UnicodeText, nullable=False)
    quote = Column(UnicodeText, nullable=False)

    def __init__(self, id, character, quote):
        self.id = id
        self.character = character
        self.quote = quote

    def __repr__(self):
        return f"id={self.ud}, character={self.character}, quote={self.quote}"


class SuperHeroQuotesDB:
    """
    The main database class, which does the following:

    * creates table for SQL database if not exists.
    * starts database connection using scoped session.
    * fetches and loads quotes from database into memory.
    * holds all methods to fetch/insert/delete quotes.

    """

    def __init__(self):
        self._dcu_quotes = []
        self._mcu_quotes = []
        self._session = self._start_session()
        self._lock = RLock()
        self.__reload_quotes()  # load quotes in memory

    @property
    def total(self):
        """
        Returns total number of quotes in database.
        """
        return len(self._dcu_quotes + self._mcu_quotes)

    @property
    def all(self):
        """
        Returns list of all quotes from database.
        """
        return self._mcu_quotes + self._dcu_quotes

    @staticmethod
    def _start_session() -> scoped_session:
        """Creates SQLAlchemy scoped session."""
        engine = create_engine(DB_URI, client_encoding="utf8")
        Base.metadata.bind = engine
        Base.metadata.create_all(engine)

        # create database tables if not exist already
        MCUQuotes.__table__.create(checkfirst=True)
        DCUQuotes.__table__.create(checkfirst=True)

        return scoped_session(sessionmaker(bind=engine, autoflush=False))

    def get_quotes(self, cat, size=10):
        """Get quotes for category with size limit."""

        cat = cat.lower()
        if cat in {"mcu", "dcu"}:
            if cat == "mcu":
                results = random.sample(self._mcu_quotes, size)
            elif cat == "dcu":
                results = random.sample(self._dcu_quotes, size)

            banner = (
                "Marvel Cinematic Universe (MCU)"
                if cat == "mcu"
                else "DC Universe (DCU)"
            )
            jsondata = {
                "StatusCode": 200,
                "Banner": banner,
                "TotalQuotes": len(results),
                "Items": results,
            }
            return jsondata

        return None

    def quote_by_id(self, qid):
        """Get specific quote with it's id"""

        all_quotes = self._mcu_quotes + self._dcu_quotes
        # skipcq: PTC-W0063
        quote = next((item for item in all_quotes if item["id"] == qid), None)
        if quote:
            return {"StatusCode": 200, "Stuff": quote}
        return None

    def ramdom_quote(self):
        """Get random quotes from both dc & marvel collection"""

        r_banner = random.choice([self._mcu_quotes, self._dcu_quotes])
        b_name = (
            "Marvel Cinematic Universe (MCU)"
            if r_banner is self._mcu_quotes
            else "DC Universe (DCU)"
        )
        return {
            "StatusCode": 200,
            "Banner": b_name,
            "Stuff": random.choice(r_banner),
        }

    def insert(self, char, quote, table):
        """Insert quotes in database."""

        with self._lock:
            table = MCUQuotes if table == "mcu" else DCUQuotes
            qid = shortuuid.uuid()
            check = self._session.query(table).get(qid)

            if not check:
                addr = table(qid, char, quote)
                self._session.add(addr)
                self._session.commit()
                return True

            self._session.close()
            return False

    def delete(self, qid, table):
        """Delete quote with the id from DB"""

        with self._lock:
            table = MCUQuotes if table == "mcu" else DCUQuotes
            curr = self._session.query(table).get(qid)
            if curr:
                self._session.delete(curr)
                self._session.commit()
                return True

            self._session.close()
            return False

    def __reload_quotes(self):
        """
        Load all quotes in memory for faster access.
        """

        try:
            mcu = self._session.query(MCUQuotes).all()
            dcu = self._session.query(DCUQuotes).all()

            for x in mcu:
                self._mcu_quotes.append(
                    {
                        "id": x.id,
                        "data": {"author": x.character, "quote": x.quote},
                    }
                )

            for x in dcu:
                self._dcu_quotes.append(
                    {
                        "id": x.id,
                        "data": {"author": x.character, "quote": x.quote},
                    }
                )
        finally:
            self._session.close()

    def __repr__(self):
        return f"{self.__class__.__name__}(total={self.total})"


QuotesAPI/__main__.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# MIT License
#
# Copyright (C) 2020-2021 Stɑrry Shivɑm
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


"""
Restful Superhero QuoteAPI written in python3 with Flask, Falsk-RESTful
and SQLAlchemy database, by (www.github.com/starry69) as fun project.

Also since i actually needed a good MCU and DCU Quotes collection
and was looking for some nice API but didn't found one so decided
to make my own, Cause why not!
"""


from flask import Flask, render_template
from flask_restful import Api, Resource, abort, reqparse
from flask_cors import CORS
from QuotesAPI.database.quotes_sql import SuperHeroQuotesDB
from QuotesAPI import ACCESS_KEY

app = Flask(
    __name__, template_folder="../templates", static_folder="../static"
)
CORS(app)  # Enable cross-domain headers for all routes
api = Api(app)
quot = SuperHeroQuotesDB()


class GetRandom(Resource):
    """
    Fetch random quotes from both mcu & dcu banner.
    """

    @staticmethod
    def get():
        return quot.ramdom_quote(), 200


class QuoteByID(Resource):
    """
    Fetch "specific" quote by it's Id
    """

    @staticmethod
    def get(quote_id):
        quote = quot.quote_by_id(quote_id)
        if quote is not None:
            return quote, 200
        abort(404, message="Can't find any quote for this id.")


# request parser for /grab endpoint.
grab_parser = reqparse.RequestParser()
grab_parser.add_argument(
    "banner",
    type=str,
    required=True,
    help="Banner of quotes to look for is required!",
)
grab_parser.add_argument(
    "size",
    type=int,
    required=False,
    help="Optionnal: total numbers of quotes to fetch",
)


class GrabCategory(Resource):
    """Get quotes for the category"""

    @staticmethod
    def get():
        args = grab_parser.parse_args()
        if args["size"] is not None:
            if args["size"] > 50:
                abort(400, message="You can only fetch max 50 quotes at time!")
            data = quot.get_quotes(args["banner"], args["size"])
        else:
            data = quot.get_quotes(args["banner"])

        if data is not None:
            return data, 200

        abort(
            404,
            message="Not found: Invalid banner provided!",
        )


# request parser for /insert endpoint.
insert_parser = reqparse.RequestParser()
insert_parser.add_argument(
    "access_key",
    type=str,
    required=True,
    help="Acess key is required to insert quote in database",
)
insert_parser.add_argument(
    "char",
    type=str,
    required=True,
    help="Charater / author of quote is required!",
)
insert_parser.add_argument(
    "quote", type=str, required=True, help="Quote to insert is required!"
)
insert_parser.add_argument(
    "table",
    type=str,
    required=True,
    help="Table name to insert quotes into is required!",
)


class InsertQuote(Resource):
    """
    Inserts quote in database availabe
    to devlopers only!
    """

    @staticmethod
    def post():
        args = insert_parser.parse_args()

        if args["access_key"] != ACCESS_KEY:
            abort(400, message="Invalid acess key!")

        if args["table"] in {"mcu", "dcu"}:
            quot.insert(
                char=args["char"], quote=args["quote"], table=args["table"]
            )
            return {"message": "Successfully inserted"}, 201

        abort(400, message="table parameter must be either 'dcu' or 'mcu'")


@app.route("/")
def index():
    """Renders homepage."""
    return render_template("index.html")


# add API resources
api.add_resource(GetRandom, "/random", endpoint="random")
api.add_resource(GrabCategory, "/grab", endpoint="grab")
api.add_resource(QuoteByID, "/quoteId/<string:quote_id>")
api.add_resource(InsertQuote, "/insert", endpoint="insert")


if __name__ == "__main__":
    app.run(debug=True, threaded=True)
